client/ayon_usd/standalone/usd/pinning/pinning_file_helper.py

```diff
@@ -0,0 +1,281 @@
+            for time in time_samples:
+                value = layer.QueryTimeSample(prop.path, time)
+                prop_data.append(value.path)
+
+    return prop_data
+
+
+def _get_prim_spec_hierarchy_external_refs(
+    prim: Sdf.PrimSpec, layer: Sdf.Layer
+) -> List[str]:
+    file_list: List[str] = _get_prim_prop_data(prim, layer)
+
+    for child_prim in prim.nameChildren:
+        file_list.extend(_get_prim_spec_hierarchy_external_refs(child_prim, layer))
+    return file_list
+
+
+def _remove_sdf_args(ref: str) -> str:
+    uri = re.sub(re.compile(r":SDF_FORMAT_ARGS.*$"), "", ref)
+    return uri
+
+
+def _resolve_udim(udim_identifier: str, layer: Sdf.Layer) -> Dict[str, str]:
+    udim_data = {}
+    resolved_udims = UsdShade.UdimUtils.ResolveUdimTilePaths(
+        str(udim_identifier), layer
+    )
+
+    for resolved_udim_path in resolved_udims:
+        udim_replaced_identifier = udim_identifier.replace(
+            "<UDIM>",
+            resolved_udim_path[1],
+        )
+
+        udim_data[udim_replaced_identifier] = resolved_udim_path[0]
+    return udim_data
+
+
+# TODO refactor so that this function has a gather block and a write block currently all individual blocks write to the identifier_to_path_dict and that makes sanitizing the data hard
+def get_asset_dependencies(layer_path: str, resolver: Ar.Resolver) -> Dict[str, str]:
+    """Return mapping from all used asset identifiers to the resolved filepaths.
+
+    Recursively traverse `Sdf.Layer` and get all their asset dependencies.
+    For each asset identifier map it to its resolved filepath.
+
+    Args:
+        layer_path: Usd layer path to be taken as the root layer
+
+    Returns: Mapping from asset identifier to their resolved paths
+
+    """
+    layer_path = _remove_sdf_args(layer_path)
+    if not layer_path:
+        return {}
+    if isinstance(layer_path, Ar.ResolvedPath):
+        layer_path = layer_path.GetPathString()
+
+    identifier_to_path: Dict[str, str] = {}
+
+    resolved_layer_path: Ar.ResolvedPath = resolver.Resolve(layer_path)
+    layer: Sdf.Layer = Sdf.Layer.FindOrOpen(resolved_layer_path)
+    identifier_to_path[layer_path] = resolved_layer_path.GetPathString()
+
+    prim_spec_file_paths: List[str] = _get_prim_spec_hierarchy_external_refs(
+        layer.pseudoRoot, layer
+    )
+    for prim_spec in prim_spec_file_paths:
+        if "<UDIM>" in prim_spec:
+
+            prim_spec = _remove_sdf_args(prim_spec)
+            unresolved_udim_path = resolver.Resolve(
+                layer.ComputeAbsolutePath(prim_spec)
+            )
+            identifier_to_path[prim_spec] = unresolved_udim_path.GetPathString()
+
+            udim_data = _resolve_udim(prim_spec, layer)
+            identifier_to_path.update(udim_data)
+            continue
+
+        prim_spec = _remove_sdf_args(prim_spec)
+
+        identifier_to_path[prim_spec] = resolver.Resolve(
+            layer.ComputeAbsolutePath(prim_spec)
+        ).GetPathString()
```

Comment:
The `prim_spec` variable name here really got to me. It's not a prim spec, but a filepath or even an asset identifier.

Also, I have a feeling this does the same thing but is shorter and more readable:
```suggestion
    for identifier in prim_spec_file_paths:
        identifier = _remove_sdf_args(identifier)
        resolved_path = resolver.Resolve(
            layer.ComputeAbsolutePath(identifier)
        )
        identifier_to_path[identifier] = resolved_path.GetPathString()
        
        if "<UDIM>" in filepath:
            # Include all tiles/paths of the UDIM
            udim_data = _resolve_udim(identifier, layer)
            identifier_to_path.update(udim_data)
```

Reply:
Actually thinking about it now - is it correct that we store the `identifier` in the dict with the SDF arguments removed? Like, is that a correct entry for the pinning file? Or should, in the pinning file the key still have the SDF arguments?

Reply:
TLDR: the key must match exactly what the resolver gets as input. This should be without SDF arguments, but:

We know by now that USD does not Strip the SDF args in every implementation. 

This means that as long as we don't solve the SDF args problem on the resolver, we will get into a situation where the pinning file generation needs to know whether the pinning file is generated for a resolver that will run for a library that has the bug or not. 

Alternatively, we can solve the SDF arguments error in the resolver, as I proposed a while ago, and not on the server. 

I take better ideas if someone has something I forgot, not sure if I overlooked something

Reply:
I'm fine either way, but it _sounds_ like for now we are in the need to write it into the pinning file with the SDF arguments, but as soon as there's a compiled version where the resolver does **not** get the SDF arguments the pinning file wouldn't work. Which makes it sounds like, for the time being... we'll need to write the pinning file with the paths mapped using BOTH including the SDF arguments and excluding the SDF arguments?

Unless we write the resolver to always check the pinning file without the SDF arguments - then writing it without SDF arguments is correct and my 'suggestion' should be fine.